### DISCUSSION TITLE
Transforming ETL script to dagster SDA with backfilling aware of existing data

### DISCUSSION THREAD

**Discussion by EvgeniyPaskin** [dagster-io/dagster#18647] [Q&A] [answered]
Hi everyone!
First of all sorry for a noob's question. I did my homework with dagster docs (and awesome dagster.university course) but still struggle to combine all dagster concepts to a workable solution for a real task I have.

The goal: move python prefect ETL script to Dagster. To make things a little bit easier: everything run/saved locally.
The task: 
- ETL part: I have a python script which gathers daily data from REST API (requests to API are quite expensive in terms of time & $). The data for each day is saved locally as a parquet file 'data_YYYY-MM-DD.parquet'. Currently, I'm running the script from time to time, manually/visually determining last downloaded file, adjusting the start_date in the script and running it.
- Analysis part: As a next step I'm using polars and Duckdb to analyse and process full dataset (e.g. I don't process it daily, I gather everything in single large dataframe). The result of the analysis is a Polars Dataframe.

I would like to move this steps (ETL&Analysis) to Dagster to:
1. have scheduled daily updates for data (including missing data, if any days for some reason was skipped)
2. I would like the dataset to be presented as daily partitioned dataset (in dagster terminology) but need to prevent dagster backfilling all history ignoring the already existing/downloaded files.
3. I need to tell dagster to be aware of already downloaded data (preventing repetitive API requests) during materialisation - this is probably the most complicated part for me, as all my experiments (including following examples from dagster) constantly required re-materialisation.

How could the structure of such project/pipeline/SDA look like?
I would appreciate a lot even a very higl-lvl breakdown of required assets/IOManagers/OPs/Jobs/Partitions/Schedules structure in human language or a link to similar example/project. 
Frankly, at this point I can not imagine how to stitch everything from the rich dagster toolkit together.
Thanks a lot in advance!

**Comment by sryza** [comment, marked as answer]
Hi @EvgeniyPaskin - imagining you have

```python
def pull_data_from_rest_API_to_parquet(date_str):
    """Assumes date is formatted as YYYY-MM-DD"""
    ...
```
you could write

```python
@asset(partitions_def=DailyPartitionsDefinition(start_date=...))
def my_partitioned_parquet_asset(context):
    pull_data_from_rest_API_to_parquet(context.partition_key)

@asset(deps=[my_partitioned_parquet_asset])
def analysis():
    ...
```

Because it's partitioned, each time you materialize `my_partitioned_parquet_asset`, you'll select a single partition, so you'll never backfill the entire asset unless you explicitly want to.

Does that answer your question?

**Reply by EvgeniyPaskin** [reply]
Hi @sryza ! Thanks a lot for your help and suggestion. 
In such setup how can I make dagster aware about a big pile of already downloaded files/data (they cover 5 years and were downloaded before moving to dagster)?  Should I just move them manually into the target folder (I assume it is where the dagster's IOManager will be saving it) and they will be picked up automatically?

**Reply by sryza** [reply]
Here's how you can make Dagster aware of already-downloaded files: https://github.com/dagster-io/dagster/discussions/17847#discussioncomment-7523864.

**Reply by EvgeniyPaskin** [reply]
Super! Looks like exactly what I needed! 
Thank you so much!
